torch/utils/data/datapipes/iter/combining.py

```diff
@@ -55,16 +57,97 @@ def __iter__(self):
+        args:
+            datapipe: Iterable DataPipe being copied
+            num_instances: number of instances of the datapipe to create
+            buffer_size: this restricts how far ahead the leading child DataPipe
+             can read relative to the slowest child DataPipe
+    """
+    def __new__(cls, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        container = _ForkIterDataPipe(datapipe, num_instances, buffer_size)
+        return [ChildDataPipe(container, i) for i in range(num_instances)]
 
-    def __new__(cls, datapipe, instances):
-        result = []
-        buffer = list(datapipe)
-        return [IterateBuffer(buffer) for i in range(instances)]
+    def __init__(self, *arg):
+        raise Exception("__init__ called instead of __new__")
+
+
+class _ForkIterDataPipe(IterDataPipe):
+    r""" :class:`_ForkIterDataPipe`.
 
+        Container to hold instance-specific information on behalf of ForkIterDataPipe. It tracks
+        the state of its child DataPipes, maintains the buffer, and yields the next value
+        as requested by the child DataPipes.
+    """
+    def __init__(self, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        self.main_datapipe = iter(datapipe)
+        self.num_instances = num_instances
+        self.buffer: Deque = deque()
+        self.buffer_size = buffer_size
+        self.child_pointers = [0] * num_instances  # Indicate the indices of the next element to get
+        self.slowest_ptr = 0
+        self.leading_ptr = 0
+        self.end_ptr = float('inf')
+
+    def get_next(self, instance_id):
+        while self.child_pointers[instance_id] < self.end_ptr:
+            if not self.buffer or self.child_pointers[instance_id] > self.leading_ptr:
+                self.leading_ptr = self.child_pointers[instance_id]
+                if self.leading_ptr - self.slowest_ptr > self.buffer_size:
+                    raise BufferError("ForkIterDataPipe buffer overflow," +
+                                      f"buffer size {self.buffer_size} is insufficient.")
+                try:
+                    self.buffer.append(self.main_datapipe.__next__())
+                    self.child_pointers[instance_id] += 1
+                    yield self.buffer[-1]
+                except StopIteration:
+                    self.end_ptr = self.leading_ptr
+            else:  # Child pointer is slower than or equal to the leading_ptr
+                buffer_index = self.child_pointers[instance_id] - self.slowest_ptr
+                return_val = self.buffer[buffer_index]
+                self.child_pointers[instance_id] += 1
+                if self.child_pointers[instance_id] - 1 == self.slowest_ptr:
+                    new_min = min(self.child_pointers)  # Can optimize by avoiding the call to min()
+                    if self.slowest_ptr < new_min:
+                        self.slowest_ptr = new_min
+                        self.buffer.popleft()
```

Comment:
I have strong feeling that code will work wrongly, as you do popleft all indexes in the buffer are getting moved, but you are not updating child_pointers

Reply:
Please add test to cover this logic

Reply:
I compute an index that accounts for the fact that `popleft()` has occurred before accessing the elements in the buffer.

`buffer_index = self.child_pointers[instance_id] - self.slowest_ptr`

I will more some tests to make sure of this.

Reply:
Ah, right. You are stacking difference in the slowest_ptr